clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorRequestManager.java

```diff
@@ -0,0 +1,284 @@
+                                     final long rebalanceTimeoutMs) {
+        Objects.requireNonNull(groupId);
+        this.time = time;
+        this.log = logContext.logger(this.getClass());
+        this.errorHandler = errorHandler;
+        this.groupId = groupId;
+        this.rebalanceTimeoutMs = rebalanceTimeoutMs;
+        this.requestTimeoutMs = config.getInt(ConsumerConfig.REQUEST_TIMEOUT_MS_CONFIG);
+        this.coordinatorRequestState = new CoordinatorRequestState(config);
+    }
+
+    CoordinatorRequestManager(final Time time,
+                              final LogContext logContext,
+                              final ErrorEventHandler errorHandler,
+                              final String groupId,
+                              final long rebalanceTimeoutMs,
+                              final long requestTimeoutMs,
+                              final CoordinatorRequestState coordinatorRequestState) {
+        Objects.requireNonNull(groupId);
+        this.time = time;
+        this.log = logContext.logger(this.getClass());
+        this.errorHandler = errorHandler;
+        this.groupId = groupId;
+        this.rebalanceTimeoutMs = rebalanceTimeoutMs;
+        this.requestTimeoutMs = requestTimeoutMs;
+        this.coordinatorRequestState = coordinatorRequestState;
+    }
+
+    @Override
+    public NetworkClientDelegate.PollResult poll(final long currentTimeMs) {
+        if (coordinatorRequestState.canSendRequest(currentTimeMs)) {
+            NetworkClientDelegate.UnsentRequest request = makeFindCoordinatorRequest(currentTimeMs);
+            return new NetworkClientDelegate.PollResult(-1, Collections.singletonList(request));
+        }
+
+        return new NetworkClientDelegate.PollResult(
+                coordinatorRequestState.remainingBackoffMs(currentTimeMs),
+                new ArrayList<>());
+    }
+
+    private NetworkClientDelegate.UnsentRequest makeFindCoordinatorRequest(final long currentTimeMs) {
+        coordinatorRequestState.updateLastSend(currentTimeMs);
+        FindCoordinatorRequestData data = new FindCoordinatorRequestData()
+                .setKeyType(FindCoordinatorRequest.CoordinatorType.GROUP.id())
+                .setKey(this.groupId);
+        return NetworkClientDelegate.UnsentRequest.makeUnsentRequest(
+                this.time.timer(requestTimeoutMs),
+                new FindCoordinatorRequest.Builder(data),
+                new FindCoordinatorRequestHandler());
+    }
+
+    /**
+     * Mark the current coordinator null and return the old coordinator. Return an empty Optional
+     * if the current coordinator is unknown.
+     * @param cause why the coordinator is marked unknown
+     * @return Optional coordinator node that can be null.
+     */
+    protected void markCoordinatorUnknown(final String cause, final long currentTimeMs) {
+        if (this.coordinator != null) {
+            log.info("Group coordinator {} is unavailable or invalid due to cause: {}. "
```

Comment:
used to be log.info, maybe log.warn is better?

Reply:
Info seems fine. It is expected for the coordinator to change over time.